File: SCT2/trunk/test-plugins/org.yakindu.sct.generator.java.test/test-gen/org/yakindu/sct/generator/java/test/GuardedEntryTest.java
/**
 * Copyright (c) 2013 committers of YAKINDU and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     committers of YAKINDU - initial API and implementation
 */

package org.yakindu.sct.generator.java.test;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.*;
import org.yakindu.scr.guardedentry.GuardedEntryStatemachine;
import org.yakindu.scr.guardedentry.GuardedEntryStatemachine.State;
/**
 *  Unit TestCase for GuardedEntry
 */
@SuppressWarnings("all")
public class GuardedEntryTest {

	private GuardedEntryStatemachine statemachine;

	@Before
	public void setUp() {
		statemachine = new GuardedEntryStatemachine();
		statemachine.init();
	}

	@After
	public void tearDown() {
		statemachine = null;
	}

	@Test
	public void testEntryNotTakenOnStatechartEnter() {
		assertTrue(statemachine.getGuard() == false);
		statemachine.enter();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		assertTrue(statemachine.getDone() == false);
	}
	@Test
	public void testEntryTakenOnStatechartEnter() {
		statemachine.setGuard(true);
		statemachine.enter();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		assertTrue(statemachine.getDone() == true);
	}
	@Test
	public void testEntryTakenInTransition() {
		statemachine.enter();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		statemachine.raiseE();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_B));
		statemachine.setGuard(true);
		statemachine.setDone(false);
		statemachine.raiseE();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		assertTrue(statemachine.getDone());
	}
	@Test
	public void testEntryNotTakenInTransition() {
		statemachine.enter();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		statemachine.raiseE();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_B));
		statemachine.setGuard(false);
		statemachine.setDone(false);
		statemachine.raiseE();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		assertTrue(!statemachine.getDone());
	}
}

File: test-plugins/org.yakindu.sct.generator.cpp.test/gtests/GuardedEntryTest/GuardedEntryTest.cc
/**
* Copyright (c) 2017 committers of YAKINDU and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*     committers of YAKINDU - initial API and implementation
*/
#include <string>
#include "gtest/gtest.h"
#include "GuardedEntry.h"
#include "sc_types.h"
GuardedEntry* statechart;

class StatemachineTest : public ::testing::Test{
	protected:
	virtual void SetUp() {
		statechart = new GuardedEntry();
		statechart->init();
	}
	virtual void TearDown() {
		delete statechart;
	}
};

void initEntryInTransition(bool guardVar, bool doneVar){
	statechart->enter();
	EXPECT_TRUE(statechart->isStateActive(GuardedEntry::main_region_A));
	statechart->raise_e();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isStateActive(GuardedEntry::main_region_B));
	statechart->getDefaultSCI()->set_guard(guardVar);
	statechart->getDefaultSCI()->set_done(doneVar);
	statechart->raise_e();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isStateActive(GuardedEntry::main_region_A));
}

TEST_F(StatemachineTest, EntryNotTakenOnStatechartEnter) {
	
	
	EXPECT_TRUE(statechart->getDefaultSCI()->get_guard()== false);
	
	statechart->enter();
	
	EXPECT_TRUE(statechart->isStateActive(GuardedEntry::main_region_A));
	
	EXPECT_TRUE(statechart->getDefaultSCI()->get_done()== false);
	
}
TEST_F(StatemachineTest, EntryTakenOnStatechartEnter) {
	
	
	statechart->getDefaultSCI()->set_guard(true);
	
	statechart->enter();
	
	EXPECT_TRUE(statechart->isStateActive(GuardedEntry::main_region_A));
	
	EXPECT_TRUE(statechart->getDefaultSCI()->get_done()== true);
	
}
TEST_F(StatemachineTest, EntryTakenInTransition) {
	
	
	initEntryInTransition(true,false);
	
	EXPECT_TRUE(statechart->getDefaultSCI()->get_done());
	
}
TEST_F(StatemachineTest, EntryNotTakenInTransition) {
	
	
	initEntryInTransition(false,false);
	
	EXPECT_TRUE(!statechart->getDefaultSCI()->get_done());
	
}
